### PR TITLE
Only produce .xz tarballs on CI

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR adds a `./configure` option to choose which tarball compression formats to produce, and changes our CI configuration to only produce `.xz` tarballs. The release process will then recompress everything into `.gz` when producing a release.

This will drastically reduce our storage costs for CI artifacts, as we'd stop storing the same data twice. **Stable, beta and nightly releases will not be affected by this at all.**

Before landing this we'll need to increase the VM size of our release process, to recompress everything in a reasonable amount of time.
r? @Mark-Simulacrum

### CODE REVIEW

File: src/bootstrap/tarball.rs [hunk 1]
```diff
@@ -288,11 +288,24 @@ impl<'a> Tarball<'a> {
...
-        crate::dist::distdir(self.builder).join(format!("{}.tar.gz", package_name))
+        // Use either the first compression format defined, or "gz" as the default.
+        let ext = self
+            .builder
+            .config
+            .dist_compression_formats
+            .as_ref()
+            .and_then(|formats| formats.get(0))
+            .map(|s| s.as_str())
+            .unwrap_or("gz");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/tarball.rs:307]: hm, I think we should probably panic if the list is defined but empty, that seems like someone is doing the wrong thing.

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with nit fixed

File: src/bootstrap/tarball.rs (hunk 1 above)
Comment (pietroalbini) [reply src/bootstrap/tarball.rs:308]: Added an assert to ensure the list is not empty.

**Comment by m-ou-se** [comment]
This failed in https://github.com/rust-lang/rust/pull/80496: https://github.com/rust-lang-ci/rust/runs/1624327931

```
gzip: stdin: not in gzip format

tar: Child returned status 1
tar: Error is not recoverable: exiting now

command did not execute successfully: "tar" "-xzf" "/checkout/obj/build/dist/rustc-nightly-src.tar.xz" "--strip-components=1"
expected success, got: exit code: 2
```

@bors r-

**Pull request merged**